### PR TITLE
Display maximum number of comments to page and delete comments

### PR DESCRIPTION
**Pull request by emilyzhaoo**
Servlet is able to limit the number of comments it returns to page (hard-coded).
Comments displayed can be deleted via home page and is deleted from DataStore.
In the process of implementing ability to have user select how many comments to display. (not ready)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Task.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2019 Google LLC
```
Comment (EkkiD) [inline portfolio/src/main/java/com/google/sps/data/Task.java]: nit: Let's do a follow-up Pull Request to renumber all of these headers to say '2020'

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -26,21 +26,55 @@
...
+    Query query = new Query("Task");    
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+ 
+
+    // Set query limit to control maximum number of comments
+    List<Entity> sentence = results.asList(FetchOptions.Builder.withLimit(2));
+//results.asIterable()
```
Comment (EkkiD) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: if this is dead code, please remove it before merging

**Comment by emilyzhaoo** [comment]
Oops! the initial commit is what I wanted the PR on, the second commit "Error with user input" was to commit changes so a tech helper could see the code!

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -26,21 +26,58 @@
...
-    response.getWriter().println("<h1>Hello world!</h1>");
+
+    // Create a new query
+    Query query = new Query("Task");    
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Get query string
+    //int limit = Integer.parseInt(request.getQueryString());
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: remove unused code.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -26,21 +26,58 @@
...
+    Query query = new Query("Task");    
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Get query string
+    //int limit = Integer.parseInt(request.getQueryString()); 
+
+    // Set query limit to control maximum number of comments
+    List<Entity> sentence = results.asList(FetchOptions.Builder.withLimit(quantity));
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Why is a list of Entity a |sentence|?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -26,21 +26,58 @@
...
+    List<Task> tasks = new ArrayList<>();
+    for (Entity entity : sentence) {
+        long id = entity.getKey().getId();
+        String animal = (String) entity.getProperty("animal");
+        String verb = (String) entity.getProperty("verb");
+        String adj = (String) entity.getProperty("adj");
+
+        Task task = new Task(id, animal, verb, adj);
+        tasks.add(task);
+
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: remove extra lines.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -26,21 +26,58 @@
...
+    response.getWriter().println(json);
   }
-**/ 
 
    @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    quantity = Integer.parseInt(request.getParameter("quantity"));
+    System.out.println(quantity);
+
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: remove extra line.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -26,21 +26,58 @@
...
+    String json = new Gson().toJson(tasks);
+    response.getWriter().println(json);
   }
-**/ 
 
    @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    quantity = Integer.parseInt(request.getParameter("quantity"));
+    System.out.println(quantity);
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I assume you use this println to debug? Please either change the message printed to something more meaningful or remove it.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long id = Long.parseLong(request.getParameter("id"));
+
+    Key taskEntityKey = KeyFactory.createKey("Task", id);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.delete(taskEntityKey);
+  }
+}
```
Comment (azhang72) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java]: nit: extra line at end of file.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -21,10 +21,59 @@ function addRandomFact() {
...
+  });
+}
+
+/** Creates an element that represents a task, including its delete button. */
+function createTaskElement(task) {
+  const taskElement = document.createElement('li');
+  taskElement.className = 'task';
+
+  const sentence = document.createElement('span');
+  sentence.innerText = ("Last Sunday, I was " + task.verb + " and I saw this " + task.adj + " " + task.animal + ", who was also " + task.verb +".");
```
Comment (azhang72) [inline portfolio/src/main/webapp/script.js]: nit: long line (80 column limit)

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -21,10 +21,59 @@ function addRandomFact() {
...
+  fetch('/delete-data', {method: 'POST', body: params});
+}
+
+/** Gets user input from display quantity drop down menu */
+function getSelection() {
+    var quantity = document.getElementById("quantity"); 
+}
+
+/** Refreshes displayed comments when changes are made. 
+function refresh() {
```
Comment (azhang72) [inline portfolio/src/main/webapp/script.js]: Delete this function if you don't need it anymore.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (emilyzhaoo) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I chose sentence because my entities are stored as verb, animal, and adj, used to create a sentence

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -65,6 +95,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(taskEntity);
 
+    response.sendRedirect("/index.html");
```
Comment (EkkiD) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:106]: per discussion in the meeting, we can probably skip the redirect here to allow the JS to add the sentence to the DOM

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -26,21 +26,50 @@
...
+import com.google.gson.Gson;
+import com.google.sps.data.Task;
+
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-/**
+  private int quantity = 5;
```
Comment (EkkiD) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think you can provide a better UX if you store quantity only on the client side. 

Let the javascript populate the GET request with a URL parameter containing the quantity that the user selected in the UI.

File: portfolio/src/main/webapp/index.html [hunk 13]
```diff
@@ -23,24 +23,52 @@ <h1>Hi! I'm Emily Zhao!</h1>
...
+                <br/><br/>
+                <p>Enter an adjective:</p>
+                <textarea name="adj" placeholder="Adjective" cols="20" rows"2"></textarea>
+                <br/><br/>
+                <p>Enter an animal:</p>
+                <textarea name="animal" placeholder="Animal" cols="20" rows"2"></textarea>
+                <br/><br/>
+                <input type="submit" name="count"/>
+                <br/><br/>
+                <label for="quantity">Choose a Quantity to Display:</label>
```
Comment (EkkiD) [inline portfolio/src/main/webapp/index.html]: nit: move this to the "Sentences Generated" card

File: portfolio/src/main/webapp/index.html [hunk 14]
```diff
@@ -23,24 +23,52 @@ <h1>Hi! I'm Emily Zhao!</h1>
...
+                    <option value="3">3</option>
+                    <option value="4">4</option>
+                    <option value="5"selected="selected">5</option>
+                    <option value="6">6</option>
+                    <option value="7">7</option>
+                    <option value="8">8</option>
+                    <option value="9">9</option>
+                    <option value="10">10</option>
+                </select>  
+                <button type="Refresh" onclick="getSelect()">Refresh</button>
```
Comment (EkkiD) [inline portfolio/src/main/webapp/index.html]: rather than triggering a refresh here, we can trigger the loadTasks() function to fetch the tasks with the requested #

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -12,19 +12,63 @@
...
   const factContainer = document.getElementById('fact-container');
   factContainer.innerText = fact;
 }
+
+/** Fetches tasks from the server and adds them to the DOM. */
+function loadTasks() {
+  // get user selection for quantity
+  getSelect(); 
+
+  fetch('/data').then(response => response.json()).then((tasks) => {
```
Comment (EkkiD) [inline portfolio/src/main/webapp/script.js]: Since we're probably going to switch to allow this to update live, we'll want to clear the DOM before repopulating it.

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -26,53 +26,74 @@
...
 
    @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
     // Get the input from the form.
     String animal = getParameter(request, "animal", "");
     String verb = getParameter(request, "verb", "");
     String adj = getParameter(request, "adj", "");
-
+
```
Comment (EkkiD) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: remove trailing whitespace

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (EkkiD) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:106]: This comment still applies. We can have the JS reload just this section

File: portfolio/src/main/webapp/script.js [hunk 17]
```diff
@@ -12,40 +12,89 @@
...
+  fetch('/delete-data', {method: 'POST', body: params});
+}
+
+/** Gets user input from display quantity drop down menu */
+function getSelect() {
+   var element = document.getElementById("quantity");
+   return element.options[element.selectedIndex].value;
 }
 
+/** Creates a geochart and adds it to the page. */
```
Comment (EkkiD) [inline portfolio/src/main/webapp/script.js]: move this to a separate PR

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js [hunk 18]
```diff
@@ -12,40 +12,105 @@
...
+function createTaskElement(task) {
+
+    const taskElement = document.createElement('li');
+    taskElement.className = 'task';
+
+    const sentence = document.createElement('span');
+    sentence.innerText = ("Last Sunday, I was " + task.verb + " and I saw this " 
+    + task.adj + " " + task.animal + ", who was also " + task.verb +"."); 
+
+<<<<<<< HEAD
```
Comment (EkkiD) [inline portfolio/src/main/webapp/script.js]: You have a merge conflict here. Resolve before merging

**Review by EkkiD — APPROVED** [review]
LGTM
